### PR TITLE
pkg/lz4: add support for LZ4 compression/decompression

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR adds support for LZ4 compression/decompression using a package from https://github.com/lz4/lz4. A test application is provided.

Note that LZ4 requires 16k allocated on the stack to work.

### Testing procedure

- A green CI
- The test application works:

<details><summary>native</summary>

```
$ BUILD_IN_DOCKER=1 make -C tests/pkg_lz4 all test --no-print-directory 
Launching build container using image "riot/riotbuild:latest".
docker run --rm --tty --user $(id -u) -v '/usr/share/zoneinfo/Europe/Paris:/etc/localtime:ro' -v '/work/riot/RIOT:/data/riotbuild/riotbase:delegated' -e 'RIOTBASE=/data/riotbuild/riotbase' -e 'CCACHE_BASEDIR=/data/riotbuild/riotbase' -e 'BUILD_DIR=/data/riotbuild/riotbase/build' -e 'RIOTPROJECT=/data/riotbuild/riotbase' -e 'RIOTCPU=/data/riotbuild/riotbase/cpu' -e 'RIOTBOARD=/data/riotbuild/riotbase/boards' -e 'RIOTMAKE=/data/riotbuild/riotbase/makefiles'        -w '/data/riotbuild/riotbase/tests/pkg_lz4/' 'riot/riotbuild:latest' make     all
Building application "tests_pkg_lz4" for "native" with MCU "native".

[INFO] updating lz4 /data/riotbuild/riotbase/build/pkg/lz4/.pkg-state.git-downloaded
echo d44371841a2f1728a3f36839fd4b7e872d0927d3   > /data/riotbuild/riotbase/build/pkg/lz4/.pkg-state.git-downloaded
[INFO] patch lz4
"make" -C /data/riotbuild/riotbase/pkg/lz4
"make" -f /data/riotbuild/riotbase/pkg/lz4/lz4.mk -C /data/riotbuild/riotbase/build/pkg/lz4/lib
"make" -C /data/riotbuild/riotbase/boards/native
"make" -C /data/riotbuild/riotbase/boards/native/drivers
"make" -C /data/riotbuild/riotbase/core
"make" -C /data/riotbuild/riotbase/cpu/native
"make" -C /data/riotbuild/riotbase/cpu/native/periph
"make" -C /data/riotbuild/riotbase/cpu/native/stdio_native
"make" -C /data/riotbuild/riotbase/drivers
"make" -C /data/riotbuild/riotbase/drivers/periph_common
"make" -C /data/riotbuild/riotbase/sys
"make" -C /data/riotbuild/riotbase/sys/auto_init
"make" -C /data/riotbuild/riotbase/sys/test_utils/interactive_sync
   text	   data	    bss	    dec	    hex	filename
  40160	    644	  66180	 106984	  1a1e8	/data/riotbuild/riotbase/tests/pkg_lz4/bin/native/tests_pkg_lz4.elf
r
/work/riot/RIOT/tests/pkg_lz4/bin/native/tests_pkg_lz4.elf /dev/ttyACM0 
RIOT native interrupts/signals initialized.
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

Help: Press s to start test, r to print it is ready
READY
s
START
main(): This is RIOT! (Version: 2022.01-devel-455-ga5f8b-pr/pkg/lz4)
Data compressed with success (ratio: 0.79)
Data decompressed with success!
Validation done, decompressed string:
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Nisl vel pretium lectus quam id leo. Volutpat sed cras ornare arcu dui vivamus arcu felis bibendum. Velit scelerisque in dictum non consectetur a erat nam. Pretium viverra suspendisse potenti nullam ac tortor vitae purus faucibus. Tristique et egestas quis ipsum suspendisse. At imperdiet dui accumsan sit amet nulla facilisi. Pulvinar neque laoreet suspendisse interdum consectetur libero. Vulputate sapien nec sagittis aliquam malesuada bibendum arcu vitae. Sed adipiscing diam donec adipiscing tristique risus nec. Venenatis tellus in metus vulputate eu scelerisque. Id faucibus nisl tincidunt eget nullam non nisi est. Integer feugiat scelerisque varius morbi enim. Est sit amet facilisis magna etiam. Venenatis cras sed felis eget velit aliquet sagittis
```

</details>

<details><summary>nrf52832-mdk</summary>

```
$ BUILD_IN_DOCKER=1 BOARD=nrf52832-mdk make -C tests/pkg_lz4 flash test --no-print-directory 
Launching build container using image "riot/riotbuild:latest".
docker run --rm --tty --user $(id -u) -v '/usr/share/zoneinfo/Europe/Paris:/etc/localtime:ro' -v '/work/riot/RIOT:/data/riotbuild/riotbase:delegated' -e 'RIOTBASE=/data/riotbuild/riotbase' -e 'CCACHE_BASEDIR=/data/riotbuild/riotbase' -e 'BUILD_DIR=/data/riotbuild/riotbase/build' -e 'RIOTPROJECT=/data/riotbuild/riotbase' -e 'RIOTCPU=/data/riotbuild/riotbase/cpu' -e 'RIOTBOARD=/data/riotbuild/riotbase/boards' -e 'RIOTMAKE=/data/riotbuild/riotbase/makefiles'      -e 'BOARD=nrf52832-mdk'  -w '/data/riotbuild/riotbase/tests/pkg_lz4/' 'riot/riotbuild:latest' make     
Building application "tests_pkg_lz4" for "nrf52832-mdk" with MCU "nrf52".

"make" -C /data/riotbuild/riotbase/pkg/lz4
"make" -f /data/riotbuild/riotbase/pkg/lz4/lz4.mk -C /data/riotbuild/riotbase/build/pkg/lz4/lib
"make" -C /data/riotbuild/riotbase/boards/nrf52832-mdk
"make" -C /data/riotbuild/riotbase/core
"make" -C /data/riotbuild/riotbase/cpu/nrf52
"make" -C /data/riotbuild/riotbase/cpu/cortexm_common
"make" -C /data/riotbuild/riotbase/cpu/cortexm_common/periph
"make" -C /data/riotbuild/riotbase/cpu/nrf52/periph
"make" -C /data/riotbuild/riotbase/cpu/nrf52/vectors
"make" -C /data/riotbuild/riotbase/cpu/nrf5x_common
"make" -C /data/riotbuild/riotbase/cpu/nrf5x_common/periph
"make" -C /data/riotbuild/riotbase/drivers
"make" -C /data/riotbuild/riotbase/drivers/periph_common
"make" -C /data/riotbuild/riotbase/sys
"make" -C /data/riotbuild/riotbase/sys/auto_init
"make" -C /data/riotbuild/riotbase/sys/isrpipe
"make" -C /data/riotbuild/riotbase/sys/malloc_thread_safe
"make" -C /data/riotbuild/riotbase/sys/newlib_syscalls_default
"make" -C /data/riotbuild/riotbase/sys/stdio_uart
"make" -C /data/riotbuild/riotbase/sys/test_utils/interactive_sync
"make" -C /data/riotbuild/riotbase/sys/tsrb
   text	   data	    bss	    dec	    hex	filename
  15048	    128	  19792	  34968	   8898	/data/riotbuild/riotbase/tests/pkg_lz4/bin/nrf52832-mdk/tests_pkg_lz4.elf
/work/riot/RIOT/dist/tools/pyocd/pyocd.sh flash /work/riot/RIOT/tests/pkg_lz4/bin/nrf52832-mdk/tests_pkg_lz4.hex
### Flashing Target ###
0005142:INFO:load_cmd:Loading /work/riot/RIOT/tests/pkg_lz4/bin/nrf52832-mdk/tests_pkg_lz4.hex at 0x00000000
[==================================================] 100%
0006766:INFO:loader:Erased 16384 bytes (4 sectors), programmed 16384 bytes (4 pages), skipped 0 bytes (0 pages) at 9.92 kB/s
Done flashing
r
/work/riot/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200" --no-reconnect --noprefix --no-repeat-command-on-empty-line 
Twisted not available, please install it if you want to use pyterm's JSON capabilities
Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
�Help: Press s to start test, r to print it is ready
Help: Press s to start test, r to print it is ready
Help: Press s to start test, r to print it is ready
Help: Press s to start test, r to print it is ready
Help: Press s to start test, r to print it is ready
Help: Press s to start test, r to print it is ready
Help: Press s to start test, r to print it is ready
READY
s
START
main(): This is RIOT! (Version: 2022.01-devel-455-ga5f8b-pr/pkg/lz4)
Data compressed with success (ratio: 0.79)
Data decompressed with success!
Validation done, decompressed string:
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Nisl vel pretium lectus quam id leo. Volutpat sed cras ornare arcu dui vivamus arcu felis bibendum. Velit scelerisque in dictum non consectetur a erat nam. Pretium viverra suspendisse potenti nullam ac tortor vitae purus faucibus. Tristique et egestas quis ipsum suspendisse. At imperdiet dui accumsan sit amet nulla facilisi. Pulvinar neque laoreet suspendisse interdum consectetur libero. Vulputate sapien nec sagittis aliquam malesuada bibendum arcu vitae. Sed adipiscing diam donec adipiscing tristique risus nec. Venenatis tellus in metus vulputate eu scelerisque. Id faucibus nisl tincidunt eget nullam non nisi est. Integer feugiat scelerisque varius morbi enim. Est sit amet facilisis magna etiam. Venenatis cras sed felis eget velit aliquet sagittis.

```

</details>

### Issues/PRs references

None

### CODE REVIEW

File: tests/pkg_lz4/main.c [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+    const int compressed_data_size = LZ4_compress_default(src, compressed_data, src_size, BUFFER_SIZE);
+    compressed_data[compressed_data_size] = 0;
+    /* Check return_value to determine what happened. */
+    if (compressed_data_size <= 0) {
+        puts("Compression failed.");
+        return -1;
+    }
+
+    if (compressed_data_size > 0) {
```
Comment (benpicco) [inline tests/pkg_lz4/main.c:68]:
```suggestion
```

The condition is always true

File: tests/pkg_lz4/main.c [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        float compression_ratio = (compressed_data_size * 100) / src_size;
```
Comment (benpicco) [inline tests/pkg_lz4/main.c:69]: no need for float - just multiply everything by 100

File: tests/pkg_lz4/main.c [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                                "scelerisque varus morbi enim. Est sit amet "
+                                "facilisis magna etiam. Venenatis cras sed "
+                                "felis eget velit aliquet sagittis.";
+
+int main(void)
+{
+    /* Compression */
+    const unsigned src_size = strlen(src) + 1;
+
+    const int compressed_data_size = LZ4_compress_default(src, compressed_data, src_size, BUFFER_SIZE);
```
Comment (benpicco) [inline tests/pkg_lz4/main.c:60]:
```suggestion
    const int compressed_data_size = LZ4_compress_default(src, compressed_data,
                                                          src_size, BUFFER_SIZE);
```

File: tests/pkg_lz4/main.c [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        puts("Compression failed.");
+        return -1;
+    }
+
+    unsigned compression_ratio = ((unsigned)compressed_data_size * 10000) / src_size;
+    printf("Data compressed with success (ratio: %u.%u)\n",
+           compression_ratio / 10000, (compression_ratio / 100) % 100);
+
+    /* Decompression */
+    const int decompressed_size = LZ4_decompress_safe(compressed_data, regen_buffer, compressed_data_size, src_size);
```
Comment (benpicco) [inline tests/pkg_lz4/main.c:73]:
```suggestion
    const int decompressed_size = LZ4_decompress_safe(compressed_data, regen_buffer,
                                                      compressed_data_size, src_size);
```

**Review by benpicco — APPROVED** [review]

File: .murdock [hunk 5]
```diff
@@ -187,7 +187,7 @@ _greplist() {
 # get list of all app directories
 get_apps() {
     make -f makefiles/app_dirs.inc.mk info-applications \
-        | $(_greplist $APPS) | sort
+        | $(_greplist $APPS) | sort | grep lz4
```
Comment (benpicco) [inline .murdock:190]: just drop that :wink:

Comment (aabadie) [reply .murdock:190]: I'm not sure Murdock will pass, but I'll do it does the next time (should be fast)

Comment (benpicco) [reply .murdock:190]: you'll still need a `Makefie.ci` - you could use the new `make  create-Makefile.ci` target for that

Comment (kaspar030) [reply .murdock:190]: BTW, file scope `APPS="tests/pkg_lz4` would also have worked.

**Pull request merged**